theano/tensor/nnet/abstract_conv.py

```diff
@@ -320,13 +340,185 @@ def conv2d_grad_wrt_weights(input,
+             any given image to upsample it by the indicated ratio
+             using bilinear interpolation.
+    """
+
+    hkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle('x', 0)
+    vkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle(0, 'x')
+    kern = hkern * vkern
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 1D filters that can
+    be applied to any image to upsample it by that ratio
+    using bilinear interpolation.
+
+    :type ratio: python integer or symbolic theano
+                 variable (Constant or Scalar)
+    : param ratio: the ratio by which an image is upsampled
+        in the 1D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 1D kernels that can be applied to
+             any given image to upsample it by the indicated ratio
+             using bilinear interpolation in one dimension.
+    """
+
+    T = theano.tensor
+    half_kern = T.arange(1, ratio + 1, dtype=theano.config.floatX)
+    kern = T.concatenate([half_kern, half_kern[-2::-1]])
+
+    if normalize:
+        kern /= float(ratio)
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        input_shape=None,
+                        use_1D_kernel=True):
+    """This function takes a stack of 2D inputs and upsamples it (in 2D space)
+    by the given ratio using bilinear interpolation with fixed (not learned)
+    kernels. When the upsampling ratio is even, the last row and colum is
+    repeated one extra time compared to the first row and column which makes
+    the upsampled tensor asymmetrical on both sides. This does not happen when
+    the upsampling ratio is odd.
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape`` that will be upsampled.
+
+    :type ratio: integer
+    : param ratio: the ratio by which the input is upsampled
+        in the 2D space.
+
+    :type input_shape: None, tuple/list of len 4 of int or
+        Constant variable.
+    :param input_shape: The shape of the input variable.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type use_1D_kernel: bool
+    :param use_1D_kernel: if set to true, row and column will
+        be upsampled seperately by 1D kernels, otherwise they are
+        upsampled together using a 2D kernel. The final result is the
+        same, only the speed can differ, given factors such as
+        upsampling ratio.
+    """
+
+    T = theano.tensor
+    bs, ch, row, col = input.shape
```

Comment:
I think I understand what you meant at the time by taking only batch_size and number_of_channels, since the last two dimensions can be taken as symbolic variables from input itself. However, the first two dims are still required. Given also what you have written here, we just need to take from the user only the first two dims.